articles/tech-lead-needed.md

```diff
@@ -0,0 +1,98 @@
+
+* In the same way, a good tech lead will act as an interpreter between the customer and the development team.
+
+* As for the architecure of the software system, a tech lead should keep the following rule of thumb in mind: Developers implement features, and implementation always is a bottom-up process. Architecture, on the other hand, is a top-down issue: Starting from the product vision, it defines a technical breakdown of the system into components and conventions that result in a uniform, intellectually manageable whole. Hence, there is a sweet spot: The tech lead needs to define the architecture down to a level that the developers can work against, but no further. If the architecture definition is too shallow (underspecified), that is, if it ends on too high an abstraction level, then the developers will struggle because they don't know what to do. If the architecture definition goes too far (overspecified), that is, if it ends on too low an abstraction level, the developers will struggle because they suffocate in constraints that forbid them to solve their problems in their own way.
+
+* Moreover, a good tech lead will be aware of the following fact: Developer work is about depth, not breadth. For example, imagine a developer who has the task of developing a robust UART driver for a new microcontroller with poorly specified peripherals. This task requires the developer to dive into the silicon for hours on end. Each distraction has terrible consequences because it interrupts the focused work, forcing the developer to sort their mind and dive into the issue from scratch. The tech lead's job, on the other hand, is about breadth, not depth. Therefore, the tech lead needs to provide guidance on the overarching aspects (what should the UART driver's interface look like, how should other components use it, how should the underlying resources be managed, et cetera). The boundary -- what can the developer provide, what does the tech lead need to supplement -- depends on the seniority of the developer, and the tech lead needs to be aware of that.
+
+* Similarly, a good tech lead will foresee the concerns that the developers might overlook in their in-depth work of implementing compenents. This includes multi-threading issues such as reentrance and locking, integration with other components, dealing with version incompabilities between components, lifecycle aspects, and so on.
+
+* Furtherwise, a good tech lead will have a sixth sense for technical risks and unexpected effort. For example, the software upgrade mechanisms and the configuration management for distributed systems are usually underestimated, and a tech lead will keep that in mind when estimating a project's cost.
+
+In all modesty, I could go on endlessly. For example, we didn't even begin to investigate interpersonal matters. However, since this is supposed to be an article rather than a textbook, we should now head on to the second key factor:
+
+## Preparing a Tech Lead
```

Comment:
The first paragraph here basically tells you the same thing as above. And while the 4 concrete steps are certainly worth mentioning, I would not spend a complete chapter on them.

I'd rather expand the "Let us explore these two factors a bit more thoroughly." above to a short paragraph about "preparing". Something like:

Experience can be gained by assigning a future tech lead additional responsiblity in small steps or by providing coaching, mentoring and networking possibilities. Let us explore these the responsibilities a bit more thoroughly.